include/extractor/restriction.hpp

```diff
@@ -5,111 +5,267 @@
 #include "util/opening_hours.hpp"
 #include "util/typedefs.hpp"
 
+#include <boost/variant.hpp>
 #include <limits>
 
 namespace osrm
 {
 namespace extractor
 {
 
-struct TurnRestriction
+// OSM offers two types of restrictions, via node and via-way restrictions. We parse both into the
+// same input container
+//
+// A restriction turning at a single node. This is the most common type of restriction:
+//
+// a - b - c
+//     |
+//     d
+//
+// ab via b to bd
+struct InputNodeRestriction
 {
-    union WayOrNode {
-        OSMNodeID_weak node;
-        OSMEdgeID_weak way;
-    };
-    WayOrNode via;
-    WayOrNode from;
-    WayOrNode to;
+    OSMEdgeID_weak from;
+    OSMNodeID_weak via;
+    OSMEdgeID_weak to;
+};
 
-    std::vector<util::OpeningHours> condition;
+// A restriction that uses a single via-way in between
+//
+// f - e - d
+//     |
+// a - b - c
+//
+// ab via be to ef -- no u turn
+struct InputWayRestriction
+{
+    OSMEdgeID_weak from;
+    OSMEdgeID_weak via;
+    OSMEdgeID_weak to;
+};
 
-    struct Bits
-    { // mostly unused
-        Bits()
-            : is_only(false), uses_via_way(false), unused2(false), unused3(false), unused4(false),
-              unused5(false), unused6(false), unused7(false)
-        {
-        }
+// Outside view of the variant
+enum class RestrictionType : std::uint8_t
+{
+    NODE_RESTRICTION,
+    WAY_RESTRICTION
+};
 
-        bool is_only : 1;
-        bool uses_via_way : 1;
-        bool unused2 : 1;
-        bool unused3 : 1;
-        bool unused4 : 1;
-        bool unused5 : 1;
-        bool unused6 : 1;
-        bool unused7 : 1;
-    } flags;
+namespace restriction_details
+{
+struct Bits
+{ // mostly unused, initialised to false by default
+    Bits() : is_only(false), way_type(false), unused(false) {}
 
-    explicit TurnRestriction(NodeID node)
+    bool is_only : 1;
+    bool way_type : 1; // if false -> node_type
+    bool unused : 6;
+};
+
+} // namespace restriction
+
+struct InputTurnRestriction
+{
+    boost::variant<InputWayRestriction, InputNodeRestriction> node_or_way;
+    restriction_details::Bits flags;
+
+    OSMEdgeID_weak from() const
     {
-        via.node = node;
-        from.node = SPECIAL_NODEID;
-        to.node = SPECIAL_NODEID;
+        return flags.way_type ? boost::get<InputWayRestriction>(node_or_way).from
+                              : boost::get<InputNodeRestriction>(node_or_way).from;
```

Comment:
I don't understand this construct - above we're using a variant but we're also storing flags? The variant already contains a flag and knows what actual type it is - I think you can use a visitor or a checked get call here.